tests/functional/tests.py

```diff
@@ -1414,46 +1413,97 @@ def threaded_get_object(cli, bckt_name, obj_name, obj_data):
     finally:
         try:
             client.remove_bucket(bucket_name)
         except Exception as err:
             raise Exception(err)
 
+def get_policy_actions(stat):
+    actions = []
+    for s in stat:
+        action = s.get('Action')
+        if action not in actions:
+            actions.append(action)
+    # flatten  nested lists in actions
+    flattened_actions = []
+    for a in actions:
+        if isinstance(a, list):
+            for aa in a:
+                flattened_actions.append(aa)
+        else:
+            flattened_actions.append(a)
+    actions = [s.replace('s3:', '') for s in flattened_actions]
+    return actions
+
+def policy_validated(client, bucket_name, policy):
+    policy_dict = json.loads(client.get_bucket_policy(bucket_name))
+    actions = get_policy_actions(policy_dict.get('Statement'))
+    actions.sort()
+    expected_actions = get_policy_actions(policy.get('Statement'))
+    expected_actions.sort()
+    if expected_actions != actions:
+        return False
+    return True
+
 def test_set_bucket_policy_readonly(client, log_output):
-    # default value for function attribute for log_output is;
-    # log_output.function = "set_bucket_policy(bucket_name, prefix, policy_access)"
+    # default value for log_output.function attribute is;
+    # log_output.function = "set_bucket_policy(bucket_name, policy)"
 
     # Get a unique bucket_name
     log_output.args['bucket_name'] = bucket_name = generate_bucket_name()
-    log_output.args['prefix'] = prefix = ''
     try:
         client.make_bucket(bucket_name)
-        # Set read-only policy successfully.
-        client.set_bucket_policy(bucket_name, prefix, Policy.READ_ONLY)
+        # read-only policy
+        policy = {
+            "Version":"2012-10-17",
+            "Statement":[
+                {
+                "Sid":"",
+                "Effect":"Allow",
+                "Principal":{"AWS":"*"},
+                "Action":"s3:GetBucketLocation",
+                "Resource":"arn:aws:s3:::"+bucket_name
+                },
+                {
+                "Sid":"",
+                "Effect":"Allow",
+                "Principal":{"AWS":"*"},
+                "Action":"s3:ListBucket",
+                "Resource":"arn:aws:s3:::"+bucket_name
+                },
+                {
+                "Sid":"",
+                "Effect":"Allow",
+                "Principal":{"AWS":"*"},
+                "Action":"s3:GetObject",
+                "Resource":"arn:aws:s3:::"+bucket_name+"/*"
+                }
+            ]
+        }
+        # Set read-only policy
+        client.set_bucket_policy(bucket_name, json.dumps(policy))
         # Validate if the policy is set correctly
-        policy_name = client.get_bucket_policy(bucket_name)
-        if policy_name != Policy.READ_ONLY:
-            raise ValueError('Failed to set ReadOnly bucket policy: ' + policy_name)
+        if not policy_validated(client, bucket_name, policy):
```

Comment:
wouldn't a simple string compare of policy json to be uploaded and policy retrieved by get_bucket_policy validate this? Don't see a need for this helper function.

Reply:
I've seen different, sometimes duplicate `Action` and/or `Resource` entries returned from the server. So, the format of the json response is not guaranteed. That's why I've introduced the helper function.